README
$Id: README,v 1.15 2001/09/08 23:39:53 cmatsuoka Exp $

Quick notes - see docs/README.* for detailed information

Directories:

./VC++6.0		Visual C++ files
./OSX-ProjectBuilder	MacOS X files
./eVC3			PocketPC files
./doc			Documentation
./etc			Configuration files
./rulesets		Extra rulesets for different architectures
./src/core		AGI interpreter core files
./src/graphics		Graphic drivers
./src/sound		Sound drivers
./src/filesys		Filesystem access drivers

The following compilers/environments can be used to build sarien:

	Compiler	Operating System
	--------------- ---------------------------------	
	GNU C compiler	UNIX, Linux, BeOS, QNX, AmigaOS
	Cygwin/Mingw32	Win32
	Visual C++	Win32
	Embedded VC	WinCE
	MIPSpro		IRIX
	DJGPP		MS-DOS (uses Allegro)
	Watcom C	MS-DOS
	Turbo C++ 3.0	MS-DOS 16-bit
	DICE C		AmigaOS
	MPW		MacOS (in progress)


To build in UNIX platforms, run './configure; make'. ./configure --help
will show the build options. If you're using mingw32 or djgpp with a
UNIX-like shell (ksh, bash), copy the appropriate ruleset from rulesets/
to ./Rules.

To build using VC++, check the files in the appropriate directory.

The list of known bugs can be browsed at
http://sourceforge.net/tracker/?group_id=24759&atid=382570



doc/Changelog

Revision History
================

0.7.0 (20011122)
	- added QNX native sound driver (untested)
	- ported to DOS 16-bit using Turbo C++
	- started MacOS port using MPW (incomplete)
	- changed palette to use Amiga-ish colors instead of PC EGA
	- added "crc" command to the interpreter console
	- sprite blitting algorithm changed to work with hidden screens
	- fixed add.to.pic to erase and re-blit all sprites
	- reimplemented options -C (show game CRC), -L (list games)
	  and -P (picture viewer)
	- unk_xxx command names changed to official names
	- added Amiga Gold Rush! 2.05 game ID
	- added support to Amiga v3 games
	- added support to Amiga sound emulation
	- fixed input prompt in normal input mode
	- using table instead of computed values for priority bands
	- implemented set.pri.base (used in KQ4)
	- implemented echo.line (to repeat last input line with F3)
	- implemented adj.ego.move.to.x.y (for click-to-walk interface)
	- added mouse event support to the X11, SDL and Win32 drivers
	- added debug info on status line (activated with key F11)
	- implemented click-to-walk mouse interface
	- menu system rewritten to allow mouse operation
	- added experimental hi-res picture mode (switch with toggle.monitor)
	- added mouse support in the picture viewer
	- added SGML man page (use docbook-to-man to create the roff file)
	- added AGI Mouse 1.0 protocol emulation
	- merged Richard Houle's MacOS X port
	- merged Paul Hill's Amiga port
	- merged Vasyl Tsvirkunov's PocketPC port
	- added experimental support to noise channel
	- implemented picture offset according to configure.screen
	- implemented invisible ego signaling
	- fixed command get.num
	- fixed controller reset (moved to the end of the cycle)
	- fixed v3 game loading in big-endian machines
	- sound tuning improved, added chorus and envelope release
	- savegame subsystem rewritten

0.6.2 (20010826)
	- fixed cmd_clear_rect() to prevent crash with the IIgs demo
	- merged Amit Vainsencher's patch to _set_cel() to avoid crash in KQ1
	- fixed macros in the X11 and SDL drivers to address gcc 3.0 warnings
	- fixed AGI variable indirection in get.room.v (bug #443900)
	- removed preprocessor directives from printf() calls to avoid
	  errors with gcc 3.0
	- fixed drawing of priority rectangle in add.to.pic
	- fixed loop setting in AGI v3 when number of loops > 4 (fixes
	  ego moonwalking in KQ4)
	- fixed OSS sound driver to use native-endian 16-bit sound
	- inventory selection bugs fixed by Ziv Barber <ZivB@siliquent.com>
	- PgUp/PgDn/Home/End codes in console fixed by Ziv Barber
	- testing raw sound resource pointer before playing

0.6.1 (20010722)
	- fixed QNX build and Photon driver
	- updated ALSA driver to work in QNX
	- fixed SVGAlib driver compile errors
	- enabled sound support with SVGAlib driver
	- fixed game crashes on quit/restart
	- fixed screen corruption in item view
	- fixed game crashes in menu operation
	- cosmetic changes in status bar/inventory screen
	- added missing fix_position() call to reposition.to and
	  reposition.to.v implementations
	- decoupled x and y-axis test in update_position()
	- keypad '5' stops ego movement
	- spacebar can be used to dismiss dialog boxes
	- always assume that sprites with priority 15 are on land
	- disabling player control when stop.motion is called
	- added missing indirection in cmd_put() (fixed by Nick Sonneveld)
	- removed v25 setting from cmd_status() (fixed by Nick Sonneveld)
	- fixed y coordinate of ego when entering room from south

0.6.0 (20010625)
	- file layout changes
	- interpreter core redesigned
	- new sprite manager
	- graphics layer redesigned to work with new sprites
	- all ancillary buffers replaced by a single AGI screen buffer
	  containing both priority and visible data
	- sprite repositioning implemented
	- blocks implemented (fixes pimp blocking stairs in LSL1)
	- proper handling of previously unused view table entries and flags
	- fixed implementation of several AGI commands
	- preliminary savegame support
	- optimized pixel routines in the graphics drivers
	- PC EGA video controller 320x200 aspect ratio correction
	- implemented inventory and object viewing
	- AGI games can be loaded from the console
	- picture viewer temporarily removed, it will be put back in
	  a future release.
	- Many win32 driver fixes

0.5.3 (20010408) *unstable*
	- starting code cleanup
	  * reindentations & cosmetic changes
	  * removing redundant checks
	  * changing UINTx to int where appropriate
	  * reducing extern variable declarations in headers
	  * reducing number of typedefs and global variables
	- merged Jeremy Penner's Photon graphics driver
	- fixed X11 24 bpp support (console now looks right)
	- fixed X11 backspace key handling
	- made GNU makefiles the default Makefiles
	- fixed sound resources deallocation in agi_v2.c
	- added AGI specs SGML documentation; currently converting from
	  linuxdoc to DocBook
	- added Lance Ewing's AGI utilities to the bundle
	- added Demo Pack 5 to the game ID list
	- buildsystem fixes for XF86 DGA and BeOS
	- added support for BeOS 5 and SDL
	- case-insensitive matching to load AGI resource files in UNIX
	- portability fixes
	- added agi2pdb to the utilities set
	- added preliminary v4 driver (to directly load PalmOS DB files)

0.5.2 (20000512) *unstable*
	- Fixed speach bubbles bug for going outside of screen (draw_obj)
	- Added more inventory code for object viewing
	- Added Peter Hawkins' save patches
	- Added Magnus Kristiansen's status line patches

0.5.1 (20000110) *unstable*
	- added new id mechanism with external id file.
	- added -C/--crc to cli options for dumping the CRC/ID of a game
	- merged v2/v3 id lists into 1 list
	- fixed v2 driver to return the proper error code in detect_game
	- DOS driver now uses 64kb less memory than before
	- merged in experimental DJGPP/Allegro driver
	- merged in Ryan Gordon's SDL + command line patches.
	- changed some length limits in text priting, messageboxes to 30

src/core/cli.c
/*  Sarien - A Sierra AGI resource interpreter engine
 *  Copyright (C) 1999-2001 Stuart George and Claudio Matsuoka
 *  
 *  $Id: cli.c,v 1.28 2001/09/08 23:39:54 cmatsuoka Exp $
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; see docs/COPYING for further details.
 */

#include "sarien.h"

#ifdef USE_COMMAND_LINE

#include <stdio.h>
#include <string.h>
#include <getopt.h>

#include "sound.h"

extern char *optarg;


static void help (int argc, char **argv)
{
	/* printf() breaks GCC 3.0 build */
	fprintf (stdout,
"Syntax is -:- %s [switches] [gamedir]\n"
"\n"
"Where [switches] are optionally:\n"
#ifdef AGDS_SUPPORT
"  -a --agds          Enables sarien to play AGDS created games.\n"
#endif
"  -A --amiga         Forces the game to be seen as an Amiga word padded game.\n"
"  -C --crc           CRC and identify the game files and stop.\n"
#if 0
"  -c --cache-loading Cache loading, static resources remain cached in memory\n"
"                     but are only loaded as needed.\n"
#endif
#ifdef OPT_LIST_DICT
"  -d --list-dictionary\n"
"                     List dictionary words.\n"
#endif
"  -E --emulate-sound {type}\n"
"                     Emulate the sound of Sierra AGI running in different\n"
"                     computers. Valid emulations are pc, mac and amiga\n"
#if 0
"  -l -force-loading  Force loading of all volume information at start of\n"
"                     game which will give you faster play, this also implies\n"
"                     the caching.\n"
#endif
"  -L --list-games    List all the games in the ID file\n"
"  -F --full-screen   Run in full-screen mode if allowed by the graphics device\n"
"  -g --no-gfx-optimizations\n"
"                     Disable optimized graphic driver hacks (if available).\n"
#ifdef USE_HIRES
"  -H --hires {0|1}   Enable/disable experimental hi-res mode.\n"
#endif
"  -h --help          Display this help screen.\n"
#ifdef USE_MOUSE
"  -m --agimouse      AGI Mouse 1.0 compatibility mode.\n"
#endif
"  -n --no-sound      Disable sound output.\n"
#ifdef OPT_PICTURE_VIEWER
"  -p --picture-viewer\n"
"                     Interactive picture viewer.\n"
#endif
#ifdef OPT_LIST_OBJECTS
"  -o --list-objects  List objects.\n"
#endif
#ifndef __MSDOS__
"  -S --scale {num}   Window size scale (only for windowed graphics).\n"
"  -r --aspect-ratio {0|1}\n"
"                     Adjust aspect ratio to match the PC EGA 320x200 screen.\n"
#endif
"  -v --emulate-version {version}\n"
"                     Force version to emulate. Valid v2 game versions are:\n"
"                     2.089, 2.272, 2.440, 2.917, 2.936. Valid v3 games are:\n"
"                     3.002.086, 3.002.149.\n"
"  -V --version       Display version information.\n"
"  -x --no-x-shm      Disable X shared memory extension (if available).\n"
#if 0
"  -xd --dga          Use XFree86 DGA extension (if available).\n"
#endif
"\n"
"[gamedir] is optionally the directory the game is in, if no directory is\n"
"specified, the current directory is assumed.\n"
"\n"
"Example: %s -v2.272 /games/sierra/sq1\n", argv[0], argv[0]);
}


int parse_cli (int argc, char **argv)
{
	struct vstrings {
		UINT16	vers;
		UINT8	*string;
	} cmp_versions[]= {
		{0x2089, (UINT8*)"2.089"},
		{0x2272, (UINT8*)"2.272"},
		{0x2440, (UINT8*)"2.440"},
		{0x2917, (UINT8*)"2.917"},
		{0x2936, (UINT8*)"2.936"},

		{0x3086, (UINT8*)"3.002.086"},
		{0x3149, (UINT8*)"3.002.149"},
		{0x0, (UINT8*)""}
	};
	int o, optidx = 0;
#define OPTIONS "AaCDdE:FgH:hLlmnopr:skS:Vv:x"
	static struct option lopt[] = {
		{ "version",		0, 0, 'V' },
		{ "help",		0, 0, 'h' },
#ifdef AGDS_SUPPORT
		{ "agds",		0, 0, 'a' },
#endif
		{ "amiga",		0, 0, 'A'},
		{ "emulate-version",	1, 0, 'v' },
#if 0
		{ "force-loading",	0, 0, 'l' },
		{ "cache-loading",	0, 0, 'c' },
#endif
		{ "crc",		0, 0, 'C'},
		{ "list-games",		0, 0, 'L'},
#ifdef OPT_LIST_DICT
		{ "list-dictionary",	0, 0, 'd' },
#endif
		{ "debug",		0, 0, 'D' },
		{ "full-screen",	0, 0, 'F' },
#ifdef USE_HIRES
		{ "hires",		1, 0, 'H' },
#endif
		{ "no-sound",		0, 0, 'n' },
#ifdef OPT_LIST_OBJECTS
		{ "list-objects",	0, 0, 'o' },
#endif
#ifdef OPT_PICTURE_VIEWER
		{ "picture-viewer",     0, 0, 'p' },                            
#endif
		{ "emulate-sound",	1, 0, 'E' },
		{ "wait-key",		0, 0, 'k' },
		{ "no-x-shm",		0, 0, 'x' },
		{ "aspect-ratio",       1, 0, 'r' },
#ifdef USE_MOUSE
		{ "agimouse",		0, 0, 'm' },
#endif
		{ "scale",		1, 0, 'S' },
		{ "no-gfx-optimizations",0,0, 'g' }
	};

	UINT16 xc;
	UINT16 ec=err_OK;

	/* FIXME: Add support for a rc file for UNIX */

	/* Set defaults */
	memset (&opt, 0, sizeof (struct sarien_options));
	opt.gamerun = GAMERUN_RUNGAME;
	opt.scale = 2;
	opt.fixratio = TRUE;
	opt.gfxhacks = TRUE;
#ifdef USE_HIRES
	opt.hires = TRUE;
#endif
#ifdef MITSHM
	opt.mitshm = TRUE;
#endif
	opt.soundemu = SOUND_EMU_NONE;

	while ((o = getopt_long (argc, argv, OPTIONS, lopt, &optidx)) != -1) {
		switch (o) {
		case 'V':
			printf ("Compiled on " __DATE__ "; " __TIME__ "\n");
			printf ("Disable Copyprotection is ");
#ifdef DISABLE_COPYPROTECTION
			printf ("enabled\n");
#else
			printf ("disabled\n");
#endif
			printf("AGDS games and Russian/Cyrillic font is ");
#ifdef AGDS_SUPPORT
			printf("supported\n");
#else
			printf("not supported\n");
#endif
			printf("Listing Objects is ");
#ifdef OPT_LIST_OBJECTS
			printf("supported\n");
#else
			printf("not supported\n");
#endif
			printf("Dictionary Listing is ");
#ifdef OPT_LIST_DICT
			printf("enabled\n");
#else
			printf("disabled\n");
#endif
#ifdef USE_EGA_PALETTE
			printf("Using EGA colour palette\n");
#else
			printf("Using AMIGA colour palette\n");
#endif
			exit (0);
		case 'A':
			opt.amiga = TRUE;
			break;
		case 'a':
			opt.agds = TRUE;
			break;
		case 'C':
			opt.gamerun = GAMERUN_CRC;
			break;
		case 'c':
			opt.cache = TRUE;
			break;
#ifdef OPT_LIST_DICT
		case 'd':
			opt.gamerun = GAMERUN_WORDS;
			break;
#endif
		case 'E':
			if (!strcmp (optarg, "pc"))
				opt.soundemu = SOUND_EMU_PC;
			else if (!strcmp (optarg, "mac"))
				opt.soundemu = SOUND_EMU_MAC;
			else if (!strcmp (optarg, "amiga"))
				opt.soundemu = SOUND_EMU_AMIGA;
			else {
				fprintf (stderr, "Sound emulation \"%s\" is "
					"unknown\n", optarg);
				exit (0);
			}
			break;
		case 'F':
			opt.fullscreen = TRUE;
			break;
		case 'f':
			opt.forceload = TRUE;
			opt.cache = TRUE;
			break;
		case 'L':
			opt.gamerun = GAMERUN_GAMES;
			break;
		case 'g':
			opt.gfxhacks = FALSE;
			break;
#ifdef USE_HIRES
		case 'H':
			opt.hires = strtoul (optarg, NULL, 0);
			break;
#endif
#ifdef USE_MOUSE
		case 'm':
			opt.agimouse = TRUE;
			break;
#endif
		case 'n':
			opt.nosound = TRUE;
			break;
#ifdef OPT_LIST_OBJECTS
		case 'o':
			opt.gamerun = GAMERUN_OBJECTS;
			break;
#endif
#ifdef OPT_PICTURE_VIEWER
		case 'p':
			opt.gamerun = GAMERUN_PICVIEW;
			break;
#endif
#ifndef __MSDOS__
		case 'r':
			opt.fixratio = strtoul (optarg, NULL, 0);
			break;
		case 'S':
			opt.scale = strtoul (optarg, NULL, 0);
			if (opt.scale < 1)
				opt.scale = 1;
			if (opt.scale > 4)
				opt.scale = 4;
			break;
#endif
		case 'v':
			for (xc = 0; xc != 0xfff0 && cmp_versions[xc].vers; xc++) {
				if (!strcmp ((char*)optarg, cmp_versions[xc].string)) {
					opt.emuversion = cmp_versions[xc].vers;
					xc = 0xfff0 - 1;
				}
			}
			if (!opt.emuversion) {
				printf ("Error: bad AGI version number \"%s\"\n",
					optarg);
				exit (-1);
			}
			break;
#ifdef MITSHM
		case 'x':
			opt.mitshm = FALSE;
			break;
#endif
		case 'h':
			help (argc, argv);
		default:
			exit (-1);
		}
	}

	return ec;
}

#endif /* USE_COMMAND_LINE */

